### ISSUE TITLE
Out of memory writing a large amount of JSON

### ISSUE THREAD

**Issue by ashleydavis** [philbooth/bfj#56]
Hi @philbooth 

Thanks for your streaming JSON library.

I'm writing a book for Manning called JavaScript Data Wrangling.  I'm currently on a chapter about working with large amounts of data in JavaScript. 

Through some experimentation I've figured out the maximum output that JSON.stringify can handle. I was hoping to use your library to go above this maximum amount, but instead I'm getting is an out of memory error.

Check out the code here:

https://github.com/JavaScript-Data-Wrangling/nodejs-json-test/blob/master/write-large-json-file-with-stream.js

Just run it like this:

  node write-large-json-file-with-stream.js

I'm trying to output an array of test data that contains 1,636,800 elements to JSON. For this particular test data this is the maximum that JSON.stringify can handle (because we hit the maximum string size in Node.js - which I think is a limitation of the V8 engine).

Your library worked very nicely on a small amount of data, but when I tried to write out 1,636,800 elements I got the out of memory issue. 

Can this be fixed?

Cheers
Ashley Davis

**Comment by philbooth** [comment]
> Can this be fixed?

Yep! It's has been on my to-do list for a while, I'll make some time to dig into it this weekend.

**Comment by philbooth** [comment]
Hey @ashleydavis, although I haven't fixed the leak yet, I have found it's possible to slow it down sufficiently using the `bufferLength` for the process to complete successfully. Might be worth experimenting with at your end if you have time.

For example, on my machine, if I change your call to `bfj.streamify` like so:

```js
const stream = bfj.streamify(dataToOutput, { bufferLength: 4 })
```

...it writes a 256-megabyte `output/huge-json-file.json` to disk.

I'm still working on a proper fix.

**Comment by philbooth** [comment]
@ashleydavis, I think I've fixed this in version `5.0.0`, let me know how it goes if you get a chance to try it out. Certainly your script doesn't fail for me locally any more.

Note for anyone else out there reading this, the fix required a breaking change to the API because I had to switch away from using native promises to resolve the leak. All promises returned by this library in `5.0.0` are bluebird promises.

**Comment by ashleydavis** [comment]
Hey Phil, thanks for you quick response. I'll test it soon on a very big file!

**Comment by ashleydavis** [comment]
Just thought I'd let you know I'm still going with this. I'm trying to process 28 gb of data so it's taking a while!

**Comment by philbooth** [comment]
@ashleydavis, any update on this issue? Are you still having problems?

**Comment by ashleydavis** [comment]
Sorry I haven't had a chance to go back to it yet. Hopefully will soon.

**Comment by philbooth** [comment]
Going to mark this as resolved with the fixes in `5.0.0`, but feel free to re-open it if anyone out there is still having issues.
